src/Internal/BalanceTx/BalanceTx.purs

```diff
@@ -1,936 +1,92 @@
-assignCoinsToChangeValues changeAddress adaAvailable pairsAtStart =
-  changeValuesAtStart <#> \changeValues ->
-    unsafeFromJust "assignCoinsToChangeValues" <<< Val.toValue <$> worker
-      (adaRequiredAtStart changeValues)
-      changeValues
-  where
-  worker :: BigInt -> NonEmptyArray ChangeValue -> Array Val
-  worker adaRequired changeValues = changeValues # NEArray.uncons >>> case _ of
-    { head: x, tail }
-      | Just xs <- NEA.fromArray tail
-      , adaAvailable < adaRequired && noTokens x ->
-          worker (adaRequired - x.minCoin) xs
-    _ ->
-      let
-        adaRemaining :: BigInt
-        adaRemaining = max zero (adaAvailable - adaRequired)
-
-        changeValuesForOutputCoins :: NonEmptyArray Val
-        changeValuesForOutputCoins =
-          let
-            weights = _.outputAda <$> changeValues
-          in
-            makeChangeForCoin weights adaRemaining
-
-        changeValuesWithMinCoins :: NonEmptyArray Val
-        changeValuesWithMinCoins = assignMinCoin <$> changeValues
-      in
-        NEArray.toArray $
-          NEArray.zipWith append changeValuesWithMinCoins
-            changeValuesForOutputCoins
-    where
-    noTokens :: ChangeValue -> Boolean
-    noTokens = null <<< Val.getAssets <<< _.value
-
-    assignMinCoin :: ChangeValue -> Val
-    assignMinCoin { value: (Val _ assets), minCoin } =
-      Val minCoin assets
-
-  adaRequiredAtStart :: NonEmptyArray ChangeValue -> BigInt
-  adaRequiredAtStart = sum <<< map _.minCoin
-
-  changeValuesAtStart :: BalanceTxM (NonEmptyArray ChangeValue)
-  changeValuesAtStart =
-    for pairsAtStart \(value /\ outputAda) ->
-      { value: Val.fromValue value, outputAda, minCoin: _ } <$> minCoinFor value
-
-  minCoinFor :: Value -> BalanceTxM BigInt
-  minCoinFor value = do
-    let
-      -- NOTE: Datum here doesn't matter, we deconstruct UTxO immediately anyway
-      txOutput = mkChangeOutput changeAddress Nothing value
-    coinsPerUtxoByte <- askCoinsPerUtxoUnit
-    pure $ BigNum.toBigInt $ utxoMinAdaValue coinsPerUtxoByte txOutput
-
-type ChangeValue = { value :: Val, outputAda :: BigInt, minCoin :: BigInt }
-
-mkChangeOutput :: Address -> Maybe OutputDatum -> Value -> TransactionOutput
-mkChangeOutput changeAddress datum amount = wrap
-  { address: changeAddress, amount, datum, scriptRef: Nothing }
-
---------------------------------------------------------------------------------
--- Getters for various `Value`s
---------------------------------------------------------------------------------
-
-getRequiredValue
-  :: BigInt -> UtxoMap -> TransactionBody -> Either BalanceTxError Val
-getRequiredValue miscFee utxos txBody = do
-  getInputVal utxos txBody <#> \inputValue ->
-    ( outputValue txBody <> minFeeValue txBody <> Val miscFee Map.empty
+-- | Balances an unbalanced transaction using the specified balancer
+-- | constraints.
+defaultBalancerErr :: CtlBalancer BalanceTxError
+defaultBalancerErr transaction ctx = do
```

Comment:
A nitpick:
I'd prefer a slightly different name, this has semantics of "Default balancer error" instead it should have a meaning of something like "Default balancer with error".